service/share/share.go

```diff
@@ -145,16 +146,49 @@ func (s *service) GetSharesByNamespace(ctx context.Context, root *Root, nID name
 	}
 
 	namespacedShares := make([]Share, 0)
+	nodeCh := make(chan []format.Node)
+	errCh := make(chan error)
+	wg := sync.WaitGroup{}
+
 	for _, rootCID := range rowRootCIDs {
-		nodes, err := ipld.GetLeavesByNamespace(ctx, s.dag, rootCID, nID)
-		if err != nil {
-			return nil, err
+		wg.Add(1)
+		go func(rootCID cid.Cid) {
+			nodes, err := ipld.GetLeavesByNamespace(ctx, s.dag, rootCID, nID)
+			if err != nil {
+				errCh <- err
+				return
+			}
+			nodeCh <- nodes
+		}(rootCID)
+	}
+
+	go func() {
+		wg.Wait()
+		close(nodeCh)
+		close(errCh)
+	}()
+
+	go func() {
+		for nodes := range nodeCh {
+			for _, node := range nodes {
+				namespacedShares = append(namespacedShares, node.RawData()[1:])
+			}
+			wg.Done()
 		}
+	}()
 
-		for _, node := range nodes {
-			namespacedShares = append(namespacedShares, node.RawData()[1:])
+	var fetchErr error
+	for err := range errCh {
+		if fetchErr != nil {
```

Comment:
🤔 Are you sure this line makes sense?

Reply:
Hi @liamsi , yes , you right. The condiiton should be `if fetchErr == nil` to handle the first occurred error. But I`m thinking, do we need here to handle the first error in case of multiple errors will be occurred? Or we can return the last one and in this case we do not need this condition.

Reply:
Good question. I think we could (as before) return on the first caught error. There is no need to continue fetching after an error occurred.

Reply:
Added new context to cancel all requests if error occurred.